src/doc/trpl/ffi.md

```diff
@@ -555,3 +555,60 @@ pub extern fn oh_no() -> i32 {
 # fn main() {}
 ```
 
+# Representing opaque structs
+
+Sometimes, a C library will provide you with a pointer to some kind of structure
+that you’re not supposed to know the memory layout of. These structures are called
+‘opaque’, and one way of representing them is with the `c_void` type:
+
+```ignore
+extern crate libc;
+use libc::c_void;
+
+extern "C" {
+    pub fn foo(arg: *c_void);
+}
+
+fn main() {
+}
+```
+
+This is a perfectly valid way of handling the situation. However, we can do a bit
+better. By creating a type to represent this opaque structure, we can at least get
+type safety on the Rust side. For example, say that our library provides two
+opaque structs, `Foo` and `Bar`. With the `c_void` representation:
+
+```rust
+# #![feature(libc)]
+extern crate libc;
+use libc::c_void;
+
+extern "C" {
+    pub fn foo(arg: *mut c_void);
+    pub fn bar(arg: *mut c_void);
+}
+
+fn main() {
+}
+```
+
+We could accidentally pass a `Foo` to `bar()`, or a `Bar` to `foo()`. Instead, let’s
+create our own opaque types with `enum`:
+
+```rust
+pub enum Foo {}
+pub enum Bar {}
+
+extern "C" {
+    pub fn foo(arg: *mut Foo);
+    pub fn bar(arg: *mut Bar);
+}
+
+fn main() {
+}
+```
+
+By using an `enum` with no variants, and by _not_ annotating it with `repr(C)`, we
```

Comment:
This seems to imply that in order to create an opaque type that we can't instantiate, we must not annotate it with `repr(C)` which is a bit confusing since `#[repr(C)] enum Foo {}` doesn't compile. Maybe you want something more like

"By using an `enum` with no variants, we create an opaque type that we can't instantiate. (Note that we don't annotate it with `repr(C)`, which isn't valid syntax for enums.)"